### PR TITLE
Set up associations

### PR DESCRIPTION
**Pull request by MostafaShahat2021**
## In this PR:
&check; Set-up [Rubocop](https://www.rubocop.org/)
&check; Set-up [RSpec GitHub Action](https://github.com/microverseinc/linters-config/tree/master/ruby#optionalset-up-rspec-github-action)
&check; Following GitFlow.
-⭐️Create a Classroom class.
-⭐️Create the `has-many`/`belongs-to` relationship between Classroom and Student.
-⭐️Create a Book class.
-⭐️Create a Rental class.
-⭐️Create the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(label)
+    @label = label
+    @students = []
+  end
+
+  def add_student(student)
+    @students.push(student)
+    @students.classroom = self
+  end
```
Comment (shahadat3669) [inline classroom.rb]: 👏 Good effort in implementing the method! 👏 However, it seems that when attempting to use this method, an error of `NoMethodError` arises. Let's troubleshoot and fix this issue to ensure smooth functionality:

- [x] **NoMethodError Resolution:** It appears there's a `NoMethodError` when using the method. Double-check your code and verify that the method is defined correctly, verify that `@students.classroom` is ok or put the proper variable name.

image

By addressing this error, your method will work seamlessly. Keep up the great work! 🛠️🚀

File: student.rb [hunk 2]
```diff
@@ -1,12 +1,18 @@
...
   end
 
   def play_hooky
     '¯\(ツ)/¯'
   end
+
+  def classroom=(classroom)
+    @classroom = classroom
+    @classroom.students.push(self) unless classroom.students.include?(self)
+  end
```
Comment (shahadat3669) [inline student.rb:17]: 👏 Good attempt with the method! 👏 However, when running it, an error surfaces: `NoMethodError` in `student.rb` at line 16. The error suggests an undefined method `students` for `"first":String`. Let's resolve this to ensure your code functions smoothly:

- [ ] **Undefined Method Error:** The error indicates an undefined method `students` for a string. Review your code in `classroom.rb` and ensure that the `students` method is correctly defined. Confirm that it's called on an instance of the `classroom` class and not on a string.
image

By addressing this issue, your method should work seamlessly. Keep up the good work! 🛠️🚀

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @MostafaShahat2021, 

Great job so far! 👏 You've done well, and you're almost there for the next project ⏭️, but there are a few issues 🐛 that you still need to work on.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ created the `many-to-many` relationship between `Person` and `Book` using the intermediate class `Rental`
✅ created the  `has-many-through` relationship between `Person` and `Book` using the intermediate class `Rental`



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classroom.rb (hunk 1 above)
Comment (MostafaShahat2021) [reply classroom.rb]: Hey @shahadat3669, thanks for your suggestion, I'll try to fix that 😃

File: student.rb (hunk 2 above)
Comment (MostafaShahat2021) [reply student.rb:17]: Hey @shahadat3669, thanks for your suggestion, I'll try to fix that 😃

**Review by mckent05 — APPROVED** [review]
Hi @MostafaShahat2021, 

# STATUS: APPROVED ✔️ ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations and well done! 🎉 
approved


## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👏 👏 
- Good job setting up the associations between your classes 👍 👍 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by MostafaShahat2021** [comment]
Hey @mckent05 👋, Thanks a lot :100:  for your efforts in reviewing my code , I really appreciate it 🙂

**Pull request merged**